engine/uState.ml

```diff
@@ -1030,11 +1030,29 @@ let demote_seff_univs univs uctx =
 
 let demote_global_univs env uctx =
   let env_ugraph = Environ.universes env in
-  let global_univs = UGraph.domain env_ugraph in
-  let global_constraints, _ = UGraph.constraints_of_universes env_ugraph in
-  let promoted_uctx =
-    ContextSet.(of_set global_univs |> add_constraints global_constraints) in
-  { uctx with local = ContextSet.diff uctx.local promoted_uctx }
+  let (local_univs, local_constraints) = uctx.local in
+  let mem_univ u ugraph = match UGraph.check_declared_universes ugraph (Level.Set.singleton u) with
+  | () -> true
+  | exception (UGraph.UndeclaredLevel _) -> false
+  in
+  let mem_constraints (u, r, v) ugraph =
+    if mem_univ u ugraph && mem_univ v ugraph then match r with
+    | Le -> UGraph.check_leq ugraph (Universe.make u) (Universe.make v)
+    | Lt -> UGraph.check_leq ugraph (Universe.super @@ Universe.make u) (Universe.make v)
+    | Eq -> UGraph.check_eq_level ugraph u v
+    else false
+  in
+  let fold_univs u accu =
+    if mem_univ u env_ugraph then accu else Level.Set.add u accu
+  in
+  let fold_constraints cst accu =
+    if mem_constraints cst env_ugraph then accu else Constraints.add cst accu
+  in
+  (* local_univs minus univs(env_ugraph) *)
+  let new_univs = Level.Set.fold fold_univs local_univs Level.Set.empty in
+  (* local_constraints minus constraints(env_ugraph) *)
+  let new_constraints = Constraints.fold fold_constraints local_constraints Constraints.empty in
```

Comment:
why fold when you can filter?

Reply:
No particular reason, I like fold.

Reply:
Shouldn't filter be more efficient?

Reply:
Marginally so, in practice I don't think it matters.

Reply:
Please use filter then